Src/Support/GoogleApis.Tests/Apis/Requests/PageStreamerTest.cs

```diff
@@ -0,0 +1,227 @@
+            Assert.AreEqual(resource.AllItems, actual);
+        }
+
+        // Note: this doesn't test all the possible timings of cancellation
+        [Test]
+        public async Task Cancellation()
+        {
+            var cts = new CancellationTokenSource();
+            var gatekeeperSource = new TaskCompletionSource<int>();
+            var resource = new PageStreamedResource(
+                "check",
+                new Page("x", 1, 2, 3),
+                new Page("y", 4, 5),
+                new Page(null, 6, 7));
+            resource.gateKeeper = gatekeeperSource.Task;            
+
+            var task = PageStreamedResource.PageStreamer.FetchAllAsync(new Request(resource), cts.Token);
+            // Delay the resource fetching until we've cancelled the token
+            cts.Cancel();
+            gatekeeperSource.SetResult(0);
+            // TODO: Move to Assert.ThrowsAsync when we update to an appropriate version of NUnit.
+            // Assert.That is available, but the documentation is unclear how this interacts with await.
+            try
+            {
+                await task;
+                Assert.Fail("Expected exception");
+            }
+            catch (OperationCanceledException)
+            {
+                // Expected
+            }
+        }
+
+        // Has to be public so we can use it as a parameter for test cases
+        public class PageStreamedResource
+        {
+            internal static readonly PageStreamer<int, Request, Page, string> PageStreamer =
+                new PageStreamer<int, Request, Page, string>(
+                    (request, token) => request.Token = token,
+                    page => page.NextPageToken,
+                    page => page.Items);
+
+            // Task to await when fetching pages; defaults to an already completed task,
+            // but can be set to another task to allow for more fine-grained interaction.
+            internal Task gateKeeper = TaskEx.FromResult(0);
+            internal List<Page> Pages { get; private set; }
+            internal string Name { get; set; }
+
+            internal PageStreamedResource(string name, params Page[] pages)
+            {                
+                this.Name = name;
+                this.Pages = pages.ToList();
+                for (int i = 0; i < pages.Length - 1; i++)
+                {
+                    Pages[i].NextPageToken = Pages[i + 1].CurrentPageToken;
+                }
+            }
+
+            public override string ToString()
+            {
+                return Name;
+            }
+
+            internal Page GetPage(string token)
+            {
+                // This will fail if we provide a token that doesn't exist.
+                return Pages.Single(p => p.CurrentPageToken == token);
+            }
+
+            internal async Task<Page> GetPageAsync(string token, CancellationToken cancellationToken)
+            {
+                await gateKeeper;
```

Comment:
I think we'd want to be considering very specific scenarios in order to make that worthwhile. Just adding an extra thread wouldn't prove much, IMO. Happy to do so if you can suggest specific tests - or we can do that later should we decide it's worth it.

One thing that makes me *not* particularly keen to is that we're not actually doing any threading ourself here - we'd really be testing the C# compiler's code generation for async, and I'm fairly confident in that...